pkg/deploy/controller/deploymentconfig/controller_test.go

```diff
@@ -245,3 +246,148 @@ func TestHandle_fatalError(t *testing.T) {
 		t.Fatalf("expected a fatal error, got: %v", err)
 	}
 }
+
+// TestHandle_previousCompletedDeployment ensures that an attempt to create a
+// new deployment for a config for which all the previous deployments are
+// in a terminal state (complete, failed) is successful
+func TestHandle_previousCompletedDeployment(t *testing.T) {
+	var (
+		config              *deployapi.DeploymentConfig
+		deployed            *kapi.ReplicationController
+		existingDeployments *kapi.ReplicationControllerList
+	)
+
+	controller := &DeploymentConfigController{
+		makeDeployment: func(config *deployapi.DeploymentConfig) (*kapi.ReplicationController, error) {
+			return deployutil.MakeDeployment(config, api.Codec)
+		},
+		deploymentClient: &deploymentClientImpl{
+			getDeploymentFunc: func(namespace, name string) (*kapi.ReplicationController, error) {
+				return nil, kerrors.NewNotFound("ReplicationController", name)
+			},
+			createDeploymentFunc: func(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error) {
+				deployed = deployment
+				return deployment, nil
+			},
+			listDeploymentsForConfigFunc: func(namespace, configName string) (*kapi.ReplicationControllerList, error) {
+				return existingDeployments, nil
+			},
+		},
+		recorder: &record.FakeRecorder{},
+	}
+
+	type existing struct {
+		version int
+		status  string
```

Comment:
You already convert these with `string()` in the scenario loop; make this type `DeploymentStatus` and then you can remove all the `string()` conversion in your scenario list (would look a little cleaner).